### PR TITLE
Fvk 2023 q2 gru

### PR DESCRIPTION
**Pull request by fvankrieken**
Addresses these two issues in gru qaqc
- https://github.com/NYCPlanning/db-gru-qaqc/issues/256
- https://github.com/NYCPlanning/db-gru-qaqc/issues/257

See PR [here](https://github.com/NYCPlanning/db-gru-qaqc/pull/258) for related changes to that repo

- adds github api functionality to get info on workflows and dispatch workflows
- page for GRU essentially taking place of readme in that repo. Chiefly
  - table with latest versions of source data (new functionality)
  - table with latest outputs of checks being run, now with added columns for workflow status and a button to trigger a run
  - markdown text from readme

This will need an app token added to the running environment of the app

### CODE REVIEW

**Comment by fvankrieken** [comment]
Merging this is blocked by just a couple little things in the db-gru-qaqc branch, but it's ready for review.

i deployed changes for now, check out 

https://edm-data-engineering.nycplanningdigital.com/?page=GRU+QAQC

Running locally requires generating a github access token, I'll maybe add a note to the readme since this is our only repo that has that

**Comment by fvankrieken** [comment]
Ooh running workflow from the app fails. Maybe not ready for review after all

**Comment by fvankrieken** [comment]
Interestingly, issue was on db-gru-qaqc side. I had updated the workflow file to have options for inputs rather than free text field, but this caused a 422 (unprocessable entity) error when trying to specify input for the workflow dispatch via github api

File: src/github.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    return r.json()
+
+def __get_workflow_runs_helper(url, params=None):
+    response = requests.get(url, headers=headers, params=params).json()
+    if 'workflow_runs' in response:
+        return response['workflow_runs']
+    else:
+        raise Exception(f'Error retreiving workflow runs from Github. If error persists, contact Data Engineering. Github API response: {response}')
+
+def get_workflow_runs(repo, workflow_name = None, items_per_page = None, total_items = None):
```
Comment (mbh329) [inline src/github.py:30]: Are the links in the QAQC checks under the `status` column meant to work? I just get sent back to the DE QAQC page

Comment (fvankrieken) [reply src/github.py:30]: Oh good catch - I think I just forgot to implement that. Doing it now

File: src/github.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import os
+import requests
+import streamlit as st
+
+ORG = 'NYCPlanning'
```
Comment (mbh329) [inline src/github.py:5]: I don't think you have black formatting activated as we are mixing single quotes with double quotes based on other scripts in this repo, I think if we update the devcontainer.json file with the proper extensions based on the [template repo](https://github.com/NYCPlanning/db-template-repo/blob/main/.devcontainer/devcontainer.json)

Comment (fvankrieken) [reply src/github.py:5]: Gotcha. I'll do that. @damonmcc good example for my to try trunk?

File: requirements.txt
Comment (damonmcc) [inline requirements.txt:1]: I'm curious how packages and versions changed here in `requirements.txt` without any changes to [`requirements.in`](https://github.com/NYCPlanning/data-engineering-qaqc/blob/3d9752c25e602598390cb4d989e3b00eddf24e62/requirements.in)?

File: src/github.py (hunk 2 above)
Comment (damonmcc) [reply src/github.py:5]: happy to troubleshoot getting black to do the formatting! (only on files changed in this PR lol)

Comment (fvankrieken) [reply src/github.py]: Just used black for now

File: src/github.py (hunk 1 above)
Comment (fvankrieken) [reply src/github.py]: Deployed fix

File: src/gru/gru.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+def gru():
+    import streamlit as st
+    import time
+    from .constants import readme_markdown_text
+    from .helpers import get_qaqc_runs, run_all_workflows
+    from .components import source_table, check_table
+
+    st.header("GRU QAQC")
+    st.write(
```
Comment (mbh329) [inline src/gru/gru.py:9]: Very nit-picky but you can just use one `st.write` and even more nitpicky...the code goes to triple quoted to single quoted. Obviously this has not impact on the function of the app

File: src/github.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+def parse_workflow(workflow):
+    return {
+        "status": workflow["status"],
+        "conclusion": workflow["conclusion"],
+        "timestamp": workflow["updated_at"],
+        "url": workflow["html_url"],
+    }
+
+
+def get_workflow(repo, name):
```
Comment (mbh329) [inline src/github.py:20]: Where is this function called in the code? I don't see it called in the PR, only defined

File: src/gru/gru.py (hunk 3 above)
Comment (fvankrieken) [reply src/gru/gru.py:9]: All for the nitpicks!

File: src/github.py (hunk 4 above)
Comment (fvankrieken) [reply src/github.py:20]: [here](https://github.com/NYCPlanning/data-engineering-qaqc/pull/261/files#diff-39f7e49e9fe3d72dafea43db356c5e896174bfdbb6b2a1c879ca236ef219f28eR47)

Comment (mbh329) [reply src/github.py:20]: `ctrl + f`failed me

**Review by damonmcc — APPROVED** [review]
lgtm, approved! but suggest deploying from this branch after the final commit to catch anything we may be missing since the last delpoy

**Review by mbh329 — APPROVED** [review]
lgtm

**Pull request merged**